File: src/main/java/com/itsci/mju/maebanjumpen/model/Report.java
package com.itsci.mju.maebanjumpen.model;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "report")
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = false)
public class Report {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "reportId")
    private Integer reportId;

    @Column(name = "reportTitle", nullable = false)
    private String reportTitle;

    @Column(name = "reportMessage", columnDefinition = "TEXT")
    private String reportMessage;

    @Column(name = "reportDate", nullable = false)
    private LocalDateTime reportDate;

    @Column(name = "reportStatus", nullable = false, length = 30)
    private String reportStatus;

    // ----------------- RELATIONSHIPS -----------------

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reporter_id")
    private PartyRole reporter;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hirer_id")
    private Member hirer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "housekeeper_id")
    private Member housekeeper;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "penalty_id")
    private Penalty penalty;


    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hire_id", nullable = false)
    private Hire hire;
}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/entity/Report.kt
package com.itsci.mju.maebanjumpen.entity

import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "report")
data class Report(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "reportId")
    var id: Long? = 0,

    @Column(name = "reportTitle", nullable = false)
    var reportTitle: String,

    @Column(name = "reportMessage", columnDefinition = "TEXT")
    var reportMessage: String,

    @Column(name = "reportDate", nullable = false)
    var reportDate: LocalDateTime,

    @Column(name = "reportStatus", nullable = false)
    var reportStatus: String,


    @Column(name = "reporter_id")
    var reporterId: Long? = 0,

    @Column(name = "hire_id")
    var hireId: Long? = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reporter_id", insertable = false, updatable = false)
    var reporter: PartyRole? = null,

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "penalty_id")
    var penalty: Penalty? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hire_id", insertable = false, updatable = false)
    var hire: Hire? = null
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Report) return false
        return id != null && id == other.id
    }

    override fun hashCode(): Int = id?.hashCode() ?: 0

    override fun toString(): String = "Report(reportId=$id, reportTitle='$reportTitle', reportStatus='$reportStatus')"
}

